actors/builtin/miner/miner_actor.go

```diff
@@ -872,63 +1227,51 @@ func (a Actor) checkPrecommitExpiry(rt Runtime, sectors *abi.BitField, regProof
 			depositToBurn = big.Add(depositToBurn, sector.PreCommitDeposit)
-		}
+			return nil
+		})
 
 		st.PreCommitDeposits = big.Sub(st.PreCommitDeposits, depositToBurn)
 		Assert(st.PreCommitDeposits.GreaterThanEqual(big.Zero()))
 		return nil
 	})
 
-	_, code := rt.Send(builtin.BurntFundsActorAddr, builtin.MethodSend, nil, depositToBurn)
-	builtin.RequireSuccess(rt, code, "failed to burn funds")
-	return
+	// This deposit was locked separately to pledge collateral so there's no pledge change here.
+	burnFunds(rt, depositToBurn)
 }
 
-func (a Actor) checkSectorExpiry(rt Runtime, sectors *abi.BitField) {
-	sectorNos := bitfieldToSectorNos(rt, sectors)
-
-	var st State
-	rt.State().Readonly(&st)
-	toTerminate := []abi.SectorNumber{}
-	for _, sectorNo := range sectorNos {
-		sector, found, err := st.GetSector(adt.AsStore(rt), sectorNo)
-		if err != nil {
-			rt.Abortf(exitcode.ErrIllegalState, "failed to load sector %v: %v", sectorNo, err)
-		}
-		if found && rt.CurrEpoch() >= sector.Info.Expiration {
-			toTerminate = append(toTerminate, sectorNo)
-		}
-		// Else sector has been terminated or extended.
+// TODO: red flag that this method is potentially super expensive
+func (a Actor) terminateSectors(rt Runtime, sectorNos *abi.BitField, terminationType power.SectorTermination) {
+	empty, err := sectorNos.IsEmpty()
+	if err != nil {
+		rt.Abortf(exitcode.ErrIllegalState, "failed to count sectors")
+	}
+	if empty {
+		return
 	}
 
-	a.terminateSectors(rt, toTerminate, power.SectorTerminationExpired)
-	return
-}
-
-// TODO: red flag that this method is potentially super expensive
-func (a Actor) terminateSectors(rt Runtime, sectorNos []abi.SectorNumber, terminationType power.SectorTermination) {
 	store := adt.AsStore(rt)
 	var st State
 
 	var dealIDs []abi.DealID
-	var allWeights []*power.SectorStorageWeightDesc
-	var faultedWeights []*power.SectorStorageWeightDesc
+	var allSectors []*SectorOnChainInfo
+	var faultySectors []*SectorOnChainInfo
+	var penalty abi.TokenAmount
 
 	rt.State().Transaction(&st, func() interface{} {
-
 		maxAllowedFaults, err := st.GetMaxAllowedFaults(store)
-		amountToSlash := big.Zero()
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load fault max")
 
+		// Narrow faults to just the set that are expiring, before expanding to a map.
+		faults, err := bitfield.IntersectBitField(sectorNos, st.Faults)
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load faults")
+
+		faultsMap, err := faults.AllMap(maxAllowedFaults)
 		if err != nil {
-			rt.Abortf(exitcode.SysErrInternal, "failed to get number of sectors")
-		}
-		faultsMap, err := st.FaultSet.AllMap(maxAllowedFaults)
-		if err != nil {
-			rt.Abortf(exitcode.ErrIllegalState, "too many faults")
+			rt.Abortf(exitcode.ErrIllegalState, "failed to expand faults")
 		}
 
-		for _, sectorNo := range sectorNos {
-			sector, found, err := st.GetSector(store, sectorNo)
+		_ = sectorNos.ForEach(func(sectorNo uint64) error {
```

Comment:
ignoring errors makes me ever so slightly nervous

Reply:
This particular code aborts rather than return errors, but that nervousness is well-founded so I'll change all the non-trivial cases of this to check the error anyway.